hazelcast/src/main/java/com/hazelcast/client/impl/connection/nio/ClientConnectionManagerImpl.java

```diff
@@ -185,134 +224,299 @@ protected NioNetworking initNetworking(final HazelcastClientInstanceImpl client)
+            throw rethrow(e);
+        } catch (ClientNotAllowedInClusterException e) {
+            logger.warning("Exception during initial connection to " + address + ": " + e);
+            throw e;
+        } catch (Exception e) {
+            logger.warning("Exception during initial connection to " + address + ": " + e);
             return null;
         }
-        return activeConnections.get(inetSocketAddressCache.get(target));
     }
 
-    @Override
-    public Connection getOrConnect(Address address) throws IOException {
+    private void fireLifecycleEvent(LifecycleEvent.LifecycleState state) {
+        LifecycleServiceImpl lifecycleService = (LifecycleServiceImpl) client.getLifecycleService();
+        lifecycleService.fireLifecycleEvent(state);
+    }
+
+    private boolean connectToCandidate(CandidateClusterContext context) {
+        Set<Address> triedAddresses = new HashSet<Address>();
         try {
-            ClientConnection connection = (ClientConnection) getConnection(address);
-            if (connection != null) {
-                return connection;
-            }
-            AuthenticationFuture future = triggerConnect(address);
-            return future.get();
-        } catch (Throwable e) {
-            throw rethrow(e);
+            waitStrategy.reset();
+            do {
+                Collection<Address> addresses = getPossibleMemberAddresses(context.getAddressProvider());
+                for (Address address : addresses) {
+                    checkClientActive();
+                    triedAddresses.add(address);
+
+                    Connection connection = connect(address);
+                    if (connection != null) {
+                        return checkFailoverSupport(connection);
+                    }
+                }
+
+                // If the address providers load no addresses (which seems to be possible), then the above loop is not entered
+                // and the lifecycle check is missing, hence we need to repeat the same check at this point.
+                checkClientActive();
+
+            } while (waitStrategy.sleep());
+        } catch (ClientNotAllowedInClusterException | InvalidConfigurationException e) {
+            logger.warning("Give up trying on this cluster with name: " + context.getClusterName()
+                    + " reason: " + e.getMessage());
         }
+        logger.info("Unable to connect to any address from the cluster with name: " + context.getClusterName()
+                + ". The following addresses were tried: " + triedAddresses);
+        return false;
     }
 
     @Override
-    public Connection getOrTriggerConnect(Address target) throws IOException {
-        Connection connection = getConnection(target);
-        if (connection != null) {
-            return connection;
+    public void checkInvocationAllowed() throws IOException {
+        if (state.equals(LifecycleEvent.LifecycleState.CLIENT_CONNECTED)) {
```

Comment:
retrieving volatile state value two times at different lines of this method, they may produce different values. It may be better to get it once in the beginning and use that value.